#145 Define ICache, centralize implementation in Equinox.Core

DSilence opened the pull request:
I've finally had some free time to hack something together. Here is the summary of changes:
1) Added to interfaces `ICache` interface (note that it's `async` based). Moved the `CacheEntry` type nearby, changed the Token.supersedes logic to be customizable. Added new `CacheEntryOptions` which is a more abstract way to configure the behavior of cache items (relative/absolute expiration).
2) Removed store-specific cache logic. 
3) Extracted `System.Runtime.Caching` integration into a separate project. The implementation remains the same.
4) Added `Microsoft.Extensions.Caching` implementation. Couldn't find a good way to customize the way it calculates object size (which wasn't terribly slow) so added a way to customize it via constructor parameter.
The distributed cache implementation is also pretty straightforward given that there is an abstraction here now. 
Looking forward to feedback!

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/jet/equinox?pullRequest=161) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/jet/equinox?pullRequest=161) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/jet/equinox?pullRequest=161) it.</sub>

bartelink commented:
Thanks for doing this work; from first glance it looks pretty complete.

Quick thoughts:
- It comes at a good point as #62 is in progress and is about to add another store that can definitely benefit from a cache. @rajivhost
- I'm thinking I'd like to consider overlaps with #157 and #61 before releasing (might take a run at #157 in a bit and we can see if the outcome of that helps in any way) 
- I'd definitely like a prefix on the names of the Cache impl packages - i.e. Equinox.Cache.SystemRuntime and Equinox.Caching.MicrosoftExtensions or similar naming style in alignment with how FsCodec name plugins
- 🤔 I guess we have integration tests but I'm wondering about whether there's a good way to do unit tests
- 🤔 maybe the cache interface should go somewhere other than in the core Equinox assembly (there are some StoreHelpers that are inlined into each store impl which maybe should go into an Equinox.Storage package) - the fact that Equinox.dll has 240 lines of fs including comments is important (at least to me) - not saying it needs to happen as part of this; ultimately this makes the codebase more intelligible and we can likely best refactor from there onwards
- 🤔 I'm still slightly concerned about what something other than SRC actually buys anyone as noted in #145

Probably won't get to take a deeper look today but hopefully shortly

bartelink commented:
Some follow-on points:
- having a clear interface (especially if some unit tests and/or docs can use it)
- extracting the impls from the store impl files is a win
- it's not a major problem to introduce a new package in the main repo from a CI/CD point of view

despite suggesting introducing new packages, I'm keen to
- reduce the number of packages that need to be documented and/or considered by someone doing a 101 sample
- default to embedding/hiding stuff that's not something that most folks need to worry about
- not have single user features - i.e. a Microsoft.Extensions.Caching impl would need to be something everyone needs to justify putting it in the repo (right now I can't see a reason for anyone other than you to use it; until we have proof to the contrary I'm thinking that's a good way to look at it)
- while I know it's correct from many perspectives, a tree of packages or folders as big as https://github.com/eventflow/EventFlow/tree/develop/Source is a nightmare for me - being able to simply clone `EventStore.fs` in an application to hack around a bug or add a some logging is for me an important feature of having essentially 3 files of consequence in Equinox as a whole if you get my drift

This is making me think the best approach should be along the following lines
1. extract the interface and provide the seam - probably initially into the Equinox DLL
2. you can validate by having your caching impl in a separate assembly, satisfying yourself it works
3. initially we leave the SRC impl in each store for now
4. at such time as a stronger need for an Equinox.Storage DLL presents itself, move ICache and the default impl there
5. iff there's a reason to have more that one new DLL, consider having the SRC cache impl in its own assembly
6. if there was a Equinox.Cache.SRC and >2 users of Equinox.Cache.MEC, then hosting it in the main repo would start to make sense

For now, I'd be comfortable with working step 1, but prior to taking most of the other steps, I'd be looking to:
- cut a final v2
- round out stuff like  #157 #61 
- the other consideration is that if we're to have a #62 impl arriving soon, that's likely to force and/or guide some  of the issues in steps 2-6 above

Does any of this seem reasonable to you?

bartelink commented:
I'm looking to implement #157 soon - ideally that'd be after this gets rebased and/or cherrypicked onto master now #164 is complete.
@DSilence do you see any time opening up to do that your side any time soon? Centralizing the SRC impl would make a great addition for a V2 release (esp if/when SqlStreamStreamStore support arrives), and hopefully shouldn't be _too much_ work given you've done the hard bit ;)

DSilence commented:
Agree with the points above.
While I would personally prefer extra modularity in assemblies, given that you want to keep things simple for now I'm totally fine to go with your suggestions.

I've updated the PR and rebased it on the latest master. The interface (and related base types) were moved to Core package (to a separate Cache.fs file), and implementations are made part of EventStore.fs and Cosmos.fs. My only issue with that is that those packages remain dependent on System.Runtime.Caching - even if another caching library is used, but it's something I can live with 😄.
The Microsoft.Extensions.Caching integration package was removed for now - I might end up publishing it as a separate package (or make a blog post with the implementation sample) if you don't mind.

bartelink commented:
Thanks! Given it wouldn't materially affect ones dependency tree, what would you think about moving to a single impl of Cache as `Equinox.Cache` but housed in the `Equinox.Core` package (one thing prompting this is that we're about to sprout SqlStreamStore support, another is that i can simplify the wiring in `eqxweb`) ?
Blog posts, articles, edits to DOCUMENTATION.md are all more than welcome any time of course ;)

DSilence commented:
All up for it. I'll update the pull request.

DSilence commented:
Done, please take a look.

bartelink commented on samples/Infrastructure/Storage.fs:
```diff
@@ -76,7 +76,7 @@ module Cosmos =
         let cacheStrategy =
             if cache then
-                let c = Caching.Cache("equinox-tool", sizeMb = 50)
+                let c = Equinox.Cache("equinox-tool", sizeMb = 50)
```
now this is more general, the sample can centralize the creation of the cache. Doesn't have to happen in this PR though

DSilence replied:
slightly scared to touch it in scope of this PR. Will leave it as is for now if you don't mind.

bartelink replied:
that's reasonable - small PRs are good PRs too

bartelink commented on src/Equinox.Core/Cache.fs:
```diff
@@ -0,0 +1,58 @@
+    abstract member TryGet: key: string -> Async<(StreamToken * 'state) option>
+
+namespace Equinox
+open System.Runtime.Caching
```
I like it in one file as you've done. blank lines around the opens

bartelink commented on src/Equinox.Core/Cache.fs:
```diff
@@ -0,0 +1,58 @@
+                let policy = getPolicy cacheItemOptions
+                match cache.AddOrGetExisting(key, box entry, policy) with
+                | null ->
+                    async.Return ()
```
wrap the whole method in `async {` instead unless you're sure there's a significant perf benefit

DSilence replied:
Perf hit almost nonexistant. Will wrap in `async` for better readability.

bartelink commented on src/Equinox.Core/Equinox.Core.fsproj:
```diff
@@ -28,6 +29,7 @@
 
     <PackageReference Include="FSharp.Core" Version="3.1.2.5" Condition=" '$(TargetFramework)' != 'netstandard2.0' " />
     <PackageReference Include="FSharp.Core" Version="4.3.4" Condition=" '$(TargetFramework)' == 'netstandard2.0' " />
+    <PackageReference Include="System.Runtime.Caching" Version="4.6.0" />
```
need to also have the condition too - copy from the stores and remove from them

bartelink commented on src/Equinox.Cosmos/Cosmos.fs:
```diff
@@ -874,62 +875,39 @@ type private Category<'event, 'state>(gateway : Gateway, codec : IUnionEncoder<'
-                | SyncResult.Written (token', state') ->return SyncResult.Written (intercept stream (token', state')) }
+                | SyncResult.Conflict resync -> return SyncResult.Conflict(interceptAsync resync stream)
+                | SyncResult.Written(token', state')
+                     ->
```
inconsistent indentation?

bartelink commented on src/Equinox.Cosmos/Cosmos.fs:
```diff
@@ -775,6 +775,7 @@ open Microsoft.Azure.Documents
 open Serilog
 open System
 open System.Collections.Concurrent
+open System.Runtime.Caching
```
is this still needed?

bartelink commented on src/Equinox.Cosmos/Cosmos.fs:
```diff
@@ -939,16 +917,18 @@ type private Folder<'event, 'state>
+                    match cacheItem with
+                    | None -> return batched
+                    | Some tokenAndState -> return! cached tokenAndState
+            }
```
In this file, I put the brace on prev line for conservation of vertical space (you don't have to, but don't be surprised if I move it up later!)

bartelink commented on src/Equinox.EventStore/EventStore.fs:
```diff
@@ -442,44 +442,16 @@ type private Category<'event, 'state>(context : Context, codec : FsCodec.IUnionE
         let intercept streamName tokenAndState =
-            tee streamName tokenAndState
-            tokenAndState
+            async{
```
up to prev line (same for last brace)

bartelink commented on src/Equinox.EventStore/EventStore.fs:
```diff
@@ -489,26 +461,31 @@ module Caching =
             : ICategory<'event, 'state, string> =
-        let policy = new CacheItemPolicy(SlidingExpiration = slidingExpiration)
-        let addOrUpdateSlidingExpirationCacheEntry streamName = CacheEntry >> cache.UpdateIfNewer policy (prefix + streamName)
+        let cacheEntryGenerator (initialToken: StreamToken, initialState: 'state) = new CacheEntry<'state>(initialToken, initialState, Token.supersedes)
```
perhaps mkCacheEntry or create* to align with other naming?

bartelink commented on src/Equinox.EventStore/EventStore.fs:
```diff
@@ -489,26 +461,31 @@ module Caching =
-            match cache.TryGet(prefix + streamName) with
-            | None -> batched
-            | Some (token, state) -> cached token state
+        async {
```
up and unindent

bartelink commented:
Looks great. Some minor picky comments regarding my/house style adherence added.

bartelink commented on src/Equinox.Cosmos/Cosmos.fs:
```diff
@@ -874,62 +874,37 @@ type private Category<'event, 'state>(gateway : Gateway, codec : IUnionEncoder<'
         let intercept streamName tokenAndState =
-            tee streamName tokenAndState
-            tokenAndState
+            async{
```
collapse `async {` to prev line and `}` up a line to regain 2 lines ;)

bartelink commented on src/Equinox.Cosmos/Cosmos.fs:
```diff
@@ -939,15 +912,15 @@ type private Folder<'event, 'state>
-        member __.Load containerStream (log : ILogger): Async<StreamToken * 'state> =
-            let batched = category.Load inspectUnfolds containerStream fold initial isOrigin log
+        member __.Load containerStream (log : ILogger): Async<StreamToken * 'state>  = async {
+            let! batched = category.Load inspectUnfolds containerStream fold initial isOrigin log
```
this is a behaviour change - should be return! on L922 - same in other callsite

DSilence replied:
Totally missed that 🤦‍♂. Looks like this also caused an integration test failure.

bartelink replied:
Yeah, I've been 🙈about automating the running of those for too long; Unfortunately it's not in danger of getting to the top of my TODO list for a while :(

bartelink commented on src/Equinox.Core/Equinox.Core.fsproj:
```diff
@@ -28,6 +29,8 @@
 
     <PackageReference Include="FSharp.Core" Version="3.1.2.5" Condition=" '$(TargetFramework)' != 'netstandard2.0' " />
     <PackageReference Include="FSharp.Core" Version="4.3.4" Condition=" '$(TargetFramework)' == 'netstandard2.0' " />
+    <PackageReference Include="System.Runtime.Caching" Version="4.5.0"  Condition=" '$(TargetFramework)' == 'netstandard2.0' "/>
```
pretty sure these can now be removed elsewhere as they're transitively referenced

bartelink commented:
🎉Looks perfect now; thanks for bearing with me over the course of all the iterations.
Aside from opening up the extension point, the forcing of the introduction of Equinox.Core and moving the cache impls to a central location really improves the legibility and hence approachability of the store implementation files. Merging now cc
 @rajivhost
